compiler/rustc_mir/src/interpret/validity.rs

```diff
@@ -214,6 +214,11 @@ struct ValidityVisitor<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> {
 }
 
 impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M> {
+    fn allow_uninit_and_ptr_numbers(&self) -> bool {
+        // FIXME: make this dependent on ctfe_mode and int validation flag
+        false
```

Comment:
Huh, even though `false` should make the check always happen, when I run Miri on the example code in https://github.com/rust-lang/miri/issues/1340#issue-600900312, Miri exits successfully. Why isn't the check working?

Reply:
Hm, yes, something is not working then. Are you sure you built Miri against the right rustc?

Reply:
I'll do a fresh rebuild to make sure I'm not using the wrong build artifacts.

Reply:
Ok, I did a fresh rebuild, and the error is now reported! I'm not sure what went wrong before.

Running Miri on the example code in https://github.com/rust-lang/miri/issues/1340#issue-600900312 emits this error:

```
error: Undefined Behavior: type validation failed: encountered uninitialized bytes, but expected initialized integer bytes
   --> /Users/user/rust2/library/std/src/sys/unix/mutex.rs:140:26
    |
140 | ...sult = libc::pthread_mutexattr_destroy(self.0.as_mut_ptr());
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered uninitialized bytes, but expected initialized integer bytes
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
            
    = note: inside `<std::sys::unix::mutex::PthreadMutexAttr as std::ops::Drop>::drop` at /Users/user/rust2/library/std/src/sys/unix/mutex.rs:140:26
    = note: inside `std::ptr::drop_in_place::<std::sys::unix::mutex::PthreadMutexAttr> - shim(Some(std::sys::unix::mutex::PthreadMutexAttr))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
    = note: inside `std::sys::unix::mutex::Mutex::init` at /Users/user/rust2/library/std/src/sys/unix/mutex.rs:60:5
    = note: inside `std::sys_common::mutex::MovableMutex::new` at /Users/user/rust2/library/std/src/sys_common/mutex.rs:65:18
    = note: inside `std::sync::Mutex::<()>::new` at /Users/user/rust2/library/std/src/sync/mutex.rs:215:20
    = note: inside `std::sys_common::thread_parker::generic::Parker::new` at /Users/user/rust2/library/std/src/sys_common/thread_parker/generic.rs:20:56
    = note: inside `std::thread::Thread::new` at /Users/user/rust2/library/std/src/thread/mod.rs:1080:79
    = note: inside `std::sys_common::rt::init` at /Users/user/rust2/library/std/src/sys_common/rt.rs:23:22
    = note: inside closure at /Users/user/rust2/library/std/src/rt.rs:45:42
    = note: inside `std::panicking::try::do_call::<[closure@std::rt::lang_start_internal::{closure#1}], ()>` at /Users/user/rust2/library/std/src/panicking.rs:403:40
    = note: inside `std::panicking::try::<(), [closure@std::rt::lang_start_internal::{closure#1}]>` at /Users/user/rust2/library/std/src/panicking.rs:367:19
    = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{closure#1}], ()>` at /Users/user/rust2/library/std/src/panic.rs:133:14
    = note: inside `std::rt::lang_start_internal` at /Users/user/rust2/library/std/src/rt.rs:45:5
    = note: inside `std::rt::lang_start::<()>` at /Users/user/rust2/library/std/src/rt.rs:63:17

error: aborting due to previous error
```

This error seems spurious though because IIUC Miri can't call `extern` functions.

Reply:
Miri can evaluate *specific* extern functions that it has "shims" for; that is required to evaluate almost any Rust code. The pthread_mutex functions are supported.

Looks like there might be a bug in the implementation of pthread_mutexattr_destroy, where it fails to initialize its return value?

Reply:
Hm, no, the return value is initialized here: https://github.com/rust-lang/miri/blob/76057dc518a17cc9418401d8c6dc2440bdaad947/src/shims/posix/foreign_items.rs#L277.

pthread_mutexattr_destroy is implemented by de-initializing whatever the pointer points to: https://github.com/rust-lang/miri/blob/a1233a721d0832071a2c1b1a95895cc0b924b553/src/shims/posix/sync.rs#L354. That should be correct though.

Can you get a Miri backtracte via `MIRI_BACKTRACE=1`?

Reply:
Oh I guess `mutexattr_set_kind` calls `write_scalar` and the type of the place here is still `i32` so it validates that things are initialized, and hence doing the de-init like this does not work.

Reply:
If it's still helpful, this is the full output from Miri with `MIRI_BACKTRACE=1`:

<details><summary>Output</summary>

```
An error occurred in miri:
   0: std::backtrace_rs::backtrace::libunwind::trace
             at /Users/user/rust2/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1: std::backtrace_rs::backtrace::trace_unsynchronized
             at /Users/user/rust2/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2: std::backtrace::Backtrace::create
             at /Users/user/rust2/library/std/src/backtrace.rs:327:13
   3: std::backtrace::Backtrace::force_capture
             at /Users/user/rust2/library/std/src/backtrace.rs:310:9
   4: <rustc_middle::mir::interpret::error::InterpErrorInfo as core::convert::From<rustc_middle::mir::interpret::error::InterpError>>::from
             at /Users/user/rust2/compiler/rustc_middle/src/mir/interpret/error.rs:118:53
   5: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /Users/user/rust2/library/core/src/result.rs:1915:27
   6: rustc_const_eval::interpret::validity::ValidityVisitor<M>::try_visit_primitive
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/validity.rs:534:33
   7: <rustc_const_eval::interpret::validity::ValidityVisitor<M> as rustc_const_eval::interpret::visitor::ValueVisitor<M>>::visit_value
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/validity.rs:766:12
   8: rustc_const_eval::interpret::validity::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::validate_operand_internal
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/validity.rs:933:15
   9: rustc_const_eval::interpret::validity::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::validate_operand
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/validity.rs:974:9
  10: rustc_const_eval::interpret::place::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::write_immediate
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/place.rs:637:13
  11: rustc_const_eval::interpret::place::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::write_scalar
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/place.rs:650:9
  12: miri::helpers::EvalContextExt::write_scalar_at_offset
             at ./src/helpers.rs:570:9
  13: miri::shims::posix::sync::mutexattr_set_kind
             at ./src/shims/posix/sync.rs:47:5
  14: miri::shims::posix::sync::EvalContextExt::pthread_mutexattr_destroy
             at ./src/shims/posix/sync.rs:354:9
  15: miri::shims::posix::foreign_items::EvalContextExt::emulate_foreign_item_by_name
             at ./src/shims/posix/foreign_items.rs:276:30
  16: miri::shims::foreign_items::EvalContextExt::emulate_foreign_item_by_name
             at ./src/shims/foreign_items.rs:614:45
  17: miri::shims::foreign_items::EvalContextExt::emulate_foreign_item
             at ./src/shims/foreign_items.rs:278:15
  18: miri::shims::EvalContextExt::find_mir_or_eval_fn
             at ./src/shims/mod.rs:53:20
  19: <miri::machine::Evaluator as rustc_const_eval::interpret::machine::Machine>::find_mir_or_eval_fn
             at ./src/machine.rs:443:9
  20: rustc_const_eval::interpret::terminator::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::eval_fn_call
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/terminator.rs:295:27
  21: rustc_const_eval::interpret::terminator::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::eval_terminator
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/terminator.rs:103:17
  22: rustc_const_eval::interpret::step::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::terminator
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/step.rs:313:9
  23: rustc_const_eval::interpret::step::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::step
             at /Users/user/rust2/compiler/rustc_const_eval/src/interpret/step.rs:73:9
  24: miri::eval::eval_entry::{{closure}}
             at ./src/eval.rs:291:29
  25: miri::eval::eval_entry
             at ./src/eval.rs:285:38
  26: <miri::MiriCompilerCalls as rustc_driver::Callbacks>::after_analysis::{{closure}}
             at ./src/bin/miri.rs:82:40
  27: rustc_interface::passes::QueryContext::enter::{{closure}}
             at /Users/user/rust2/compiler/rustc_interface/src/passes.rs:778:42
  28: rustc_middle::ty::context::tls::enter_context::{{closure}}
             at /Users/user/rust2/compiler/rustc_middle/src/ty/context.rs:1771:50
  29: rustc_middle::ty::context::tls::set_tlv
             at /Users/user/rust2/compiler/rustc_middle/src/ty/context.rs:1755:9
  30: rustc_middle::ty::context::tls::enter_context
             at /Users/user/rust2/compiler/rustc_middle/src/ty/context.rs:1771:9
  31: rustc_interface::passes::QueryContext::enter
             at /Users/user/rust2/compiler/rustc_interface/src/passes.rs:778:9
  32: <miri::MiriCompilerCalls as rustc_driver::Callbacks>::after_analysis
             at ./src/bin/miri.rs:59:9
  33: rustc_driver::run_compiler::{{closure}}::{{closure}}
             at /Users/user/rust2/compiler/rustc_driver/src/lib.rs:403:16
  34: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
             at /Users/user/rust2/compiler/rustc_interface/src/queries.rs:390:19
  35: rustc_driver::run_compiler::{{closure}}
             at /Users/user/rust2/compiler/rustc_driver/src/lib.rs:312:22
  36: rustc_interface::interface::create_compiler_and_run::{{closure}}
             at /Users/user/rust2/compiler/rustc_interface/src/interface.rs:209:13
  37: rustc_span::with_source_map
             at /Users/user/rust2/compiler/rustc_span/src/lib.rs:976:5
  38: rustc_interface::interface::create_compiler_and_run
             at /Users/user/rust2/compiler/rustc_interface/src/interface.rs:203:5
  39: rustc_interface::interface::run_compiler::{{closure}}
             at /Users/user/rust2/compiler/rustc_interface/src/interface.rs:225:12
  40: rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals::{{closure}}::{{closure}}
             at /Users/user/rust2/compiler/rustc_interface/src/util.rs:158:13
  41: scoped_tls::ScopedKey<T>::set
             at /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137:9
  42: rustc_span::create_session_globals_then
             at /Users/user/rust2/compiler/rustc_span/src/lib.rs:109:5
  43: rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals::{{closure}}
             at /Users/user/rust2/compiler/rustc_interface/src/util.rs:156:9
  44: rustc_interface::util::scoped_thread::{{closure}}
             at /Users/user/rust2/compiler/rustc_interface/src/util.rs:131:24
  45: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /Users/user/rust2/library/std/src/sys_common/backtrace.rs:125:18
  46: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /Users/user/rust2/library/std/src/thread/mod.rs:481:17
  47: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /Users/user/rust2/library/core/src/panic/unwind_safe.rs:271:9
  48: std::panicking::try::do_call
             at /Users/user/rust2/library/std/src/panicking.rs:403:40
  49: std::panicking::try
             at /Users/user/rust2/library/std/src/panicking.rs:367:19
  50: std::panic::catch_unwind
             at /Users/user/rust2/library/std/src/panic.rs:133:14
  51: std::thread::Builder::spawn_unchecked::{{closure}}
             at /Users/user/rust2/library/std/src/thread/mod.rs:480:30
  52: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /Users/user/rust2/library/core/src/ops/function.rs:227:5
  53: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /Users/user/rust2/library/alloc/src/boxed.rs:1638:9
  54: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /Users/user/rust2/library/alloc/src/boxed.rs:1638:9
  55: std::sys::unix::thread::Thread::new::thread_start
             at /Users/user/rust2/library/std/src/sys/unix/thread.rs:106:17
  56: __pthread_start

error: Undefined Behavior: type validation failed: encountered uninitialized bytes, but expected initialized integer bytes
   --> /Users/user/rust2/library/std/src/sys/unix/mutex.rs:140:26
    |
140 | ...sult = libc::pthread_mutexattr_destroy(self.0.as_mut_ptr());
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered uninitialized bytes, but expected initialized integer bytes
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
            
    = note: inside `<std::sys::unix::mutex::PthreadMutexAttr as std::ops::Drop>::drop` at /Users/user/rust2/library/std/src/sys/unix/mutex.rs:140:26
    = note: inside `std::ptr::drop_in_place::<std::sys::unix::mutex::PthreadMutexAttr> - shim(Some(std::sys::unix::mutex::PthreadMutexAttr))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
    = note: inside `std::sys::unix::mutex::Mutex::init` at /Users/user/rust2/library/std/src/sys/unix/mutex.rs:60:5
    = note: inside `std::sys_common::mutex::MovableMutex::new` at /Users/user/rust2/library/std/src/sys_common/mutex.rs:65:18
    = note: inside `std::sync::Mutex::<()>::new` at /Users/user/rust2/library/std/src/sync/mutex.rs:215:20
    = note: inside `std::sys_common::thread_parker::generic::Parker::new` at /Users/user/rust2/library/std/src/sys_common/thread_parker/generic.rs:20:56
    = note: inside `std::thread::Thread::new` at /Users/user/rust2/library/std/src/thread/mod.rs:1080:79
    = note: inside `std::sys_common::rt::init` at /Users/user/rust2/library/std/src/sys_common/rt.rs:23:22
    = note: inside closure at /Users/user/rust2/library/std/src/rt.rs:45:42
    = note: inside `std::panicking::try::do_call::<[closure@std::rt::lang_start_internal::{closure#1}], ()>` at /Users/user/rust2/library/std/src/panicking.rs:403:40
    = note: inside `std::panicking::try::<(), [closure@std::rt::lang_start_internal::{closure#1}]>` at /Users/user/rust2/library/std/src/panicking.rs:367:19
    = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{closure#1}], ()>` at /Users/user/rust2/library/std/src/panic.rs:133:14
    = note: inside `std::rt::lang_start_internal` at /Users/user/rust2/library/std/src/rt.rs:45:5
    = note: inside `std::rt::lang_start::<()>` at /Users/user/rust2/library/std/src/rt.rs:63:17

error: aborting due to previous error
```

</details>

Reply:
Yeah that confirms my theory.

For now I recommend you remove this line: https://github.com/rust-lang/miri/blob/a1233a721d0832071a2c1b1a95895cc0b924b553/src/shims/posix/sync.rs#L354

Reply:
Ok, I removed that line, and now the example from the issue is emitting the correct error (:tada:):

```
error: Undefined Behavior: type validation failed at .value: encountered uninitialized bytes, but expected initialized integer bytes
 --> foo.rs:2:25
  |
2 | ...safe { std::mem::MaybeUninit::<usize>::uninit().assume_init() };
  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed at .value: encountered uninitialized bytes, but expected initialized integer bytes
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
          
  = note: inside `main` at foo.rs:2:25
  = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /Users/user/rust2/library/core/src/ops/function.rs:227:5
  = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /Users/user/rust2/library/std/src/sys_common/backtrace.rs:125:18
  = note: inside closure at /Users/user/rust2/library/std/src/rt.rs:64:18
  = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /Users/user/rust2/library/core/src/ops/function.rs:259:13
  = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /Users/user/rust2/library/std/src/panicking.rs:403:40
  = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /Users/user/rust2/library/std/src/panicking.rs:367:19
  = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /Users/user/rust2/library/std/src/panic.rs:133:14
  = note: inside closure at /Users/user/rust2/library/std/src/rt.rs:46:48
  = note: inside `std::panicking::try::do_call::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /Users/user/rust2/library/std/src/panicking.rs:403:40
  = note: inside `std::panicking::try::<isize, [closure@std::rt::lang_start_internal::{closure#2}]>` at /Users/user/rust2/library/std/src/panicking.rs:367:19
  = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /Users/user/rust2/library/std/src/panic.rs:133:14
  = note: inside `std::rt::lang_start_internal` at /Users/user/rust2/library/std/src/rt.rs:46:20
  = note: inside `std::rt::lang_start::<()>` at /Users/user/rust2/library/std/src/rt.rs:63:17

error: aborting due to previous error
```

But, many tests are still failing with a somewhat similar error; this one is about `pthread_mutex_destroy`:

```
error: Undefined Behavior: type validation failed: encountered uninitialized bytes, but expected initialized integer bytes
  --> /Users/user/rust2/library/std/src/sys/unix/mutex.rs:78:17
   |
78 |         let r = libc::pthread_mutex_destroy(self.inner.get());
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered uninitialized bytes, but expected initialized integer bytes
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `std::sys::unix::mutex::Mutex::destroy` at /Users/user/rust2/library/std/src/sys/unix/mutex.rs:78:17
   = note: inside `<std::sys_common::mutex::MovableMutex as std::ops::Drop>::drop` at /Users/user/rust2/library/std/src/sys_common/mutex.rs:98:18
   = note: inside `std::ptr::drop_in_place::<std::sys_common::mutex::MovableMutex> - shim(Some(std::sys_common::mutex::MovableMutex))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::sync::Mutex<()>> - shim(Some(std::sync::Mutex<()>))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::sys_common::thread_parker::generic::Parker> - shim(Some(std::sys_common::thread_parker::generic::Parker))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::thread::Inner> - shim(Some(std::thread::Inner))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::sync::Arc::<std::thread::Inner>::drop_slow` at /Users/user/rust2/library/alloc/src/sync.rs:1065:18
   = note: inside `<std::sync::Arc<std::thread::Inner> as std::ops::Drop>::drop` at /Users/user/rust2/library/alloc/src/sync.rs:1625:13
   = note: inside `std::ptr::drop_in_place::<std::sync::Arc<std::thread::Inner>> - shim(Some(std::sync::Arc<std::thread::Inner>))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::thread::Thread> - shim(Some(std::thread::Thread))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::sys_common::thread_info::ThreadInfo> - shim(Some(std::sys_common::thread_info::ThreadInfo))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::option::Option<std::sys_common::thread_info::ThreadInfo>> - shim(Some(std::option::Option<std::sys_common::thread_info::ThreadInfo>))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::cell::UnsafeCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>> - shim(Some(std::cell::UnsafeCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>> - shim(Some(std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::ptr::drop_in_place::<std::option::Option<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>>> - shim(Some(std::option::Option<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>>))` at /Users/user/rust2/library/core/src/ptr/mod.rs:188:1
   = note: inside `std::mem::drop::<std::option::Option<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>>>` at /Users/user/rust2/library/core/src/mem/mod.rs:898:24
   = note: inside `std::thread::local::fast::destroy_value::<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>>` at /Users/user/rust2/library/std/src/thread/local.rs:655:13
   = note: inside `std::sys::unix::thread_local_dtor::register_dtor::run_dtors` at /Users/user/rust2/library/std/src/sys/unix/thread_local_dtor.rs:89:17

error: aborting due to previous error
```

Should I remove the similar lines in https://github.com/rust-lang/miri/blob/7decf8c7202663e58ccc9ad84f01abe1ac77e128/src/shims/posix/sync.rs#L483-L497 as well? Could this cause false negatives in Miri?

Reply:
Hm, if I remove those lines, I then get a similar error, but with `libc::pthread_cond_destroy`.

Reply:
Yeah, for now remove all those lines in `sync.rs`. We'll have to figure something out before landing the change, but this is a good basis for experiments.